In [24]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
import numpy as np

In [2]:
train = pd.read_csv('P:/python/SalaryData_Train(1).csv')
train.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [3]:
test = pd.read_csv('P:/python/SalaryData_Test(1).csv')
test.head()

age   workclass      education  educationno        maritalstatus  \
0   25     Private           11th            7        Never-married   
1   38     Private        HS-grad            9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm           12   Married-civ-spouse   
3   44     Private   Some-college           10   Married-civ-spouse   
4   34     Private           10th            6        Never-married   

           occupation    relationship    race    sex  capitalgain  \
0   Machine-op-inspct       Own-child   Black   Male            0   
1     Farming-fishing         Husband   White   Male            0   
2     Protective-serv         Husband   White   Male            0   
3   Machine-op-inspct         Husband   Black   Male         7688   
4       Other-service   Not-in-family   White   Male            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            50   United-States   <=50K  
2            0            40   United-States    >50K  
3            0            40   United-States    >50K  
4            0            30   United-States   <=50K

In [4]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


In [5]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15060 entries, 0 to 15059
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            15060 non-null  int64 
 1   workclass      15060 non-null  object
 2   education      15060 non-null  object
 3   educationno    15060 non-null  int64 
 4   maritalstatus  15060 non-null  object
 5   occupation     15060 non-null  object
 6   relationship   15060 non-null  object
 7   race           15060 non-null  object
 8   sex            15060 non-null  object
 9   capitalgain    15060 non-null  int64 
 10  capitalloss    15060 non-null  int64 
 11  hoursperweek   15060 non-null  int64 
 12  native         15060 non-null  object
 13  Salary         15060 non-null  object
dtypes: int64(5), object(9)
memory usage: 1.6+ MB


In [6]:
train.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [7]:
col_cat = ['workclass', 'education', 'educationno', 'maritalstatus',
           'occupation', 'relationship', 'race', 'sex', 'native', 'Salary']

In [8]:
for i in col_cat:
    print(i, train[i].value_counts())

workclass  Private             22285
 Self-emp-not-inc     2499
 Local-gov            2067
 State-gov            1279
 Self-emp-inc         1074
 Federal-gov           943
 Without-pay            14
Name: workclass, dtype: int64
education  HS-grad         9840
 Some-college    6677
 Bachelors       5044
 Masters         1627
 Assoc-voc       1307
 11th            1048
 Assoc-acdm      1008
 10th             820
 7th-8th          557
 Prof-school      542
 9th              455
 12th             377
 Doctorate        375
 5th-6th          288
 1st-4th          151
 Preschool         45
Name: education, dtype: int64
educationno 9     9840
10    6677
13    5044
14    1627
11    1307
7     1048
12    1008
6      820
4      557
15     542
5      455
8      377
16     375
3      288
2      151
1       45
Name: educationno, dtype: int64
maritalstatus  Married-civ-spouse       14065
 Never-married             9725
 Divorced                  4214
 Separated                  939
 Widowed         

In [9]:
train1 = train.copy()
test1 = test.copy()

In [10]:
train1 = train1.drop('educationno', axis = 1)
test1 = test1.drop('educationno', axis = 1)

In [11]:
train1.columns

Index(['age', 'workclass', 'education', 'maritalstatus', 'occupation',
       'relationship', 'race', 'sex', 'capitalgain', 'capitalloss',
       'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [12]:
test1.columns

Index(['age', 'workclass', 'education', 'maritalstatus', 'occupation',
       'relationship', 'race', 'sex', 'capitalgain', 'capitalloss',
       'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [13]:
train1 = train1.apply(LabelEncoder().fit_transform)
train1.head()

age  workclass  education  maritalstatus  occupation  relationship  race  \
0   22          5          9              4           0             1     4   
1   33          4          9              2           3             0     4   
2   21          2         11              0           5             1     4   
3   36          2          1              2           5             0     2   
4   11          2          9              2           9             5     2   

   sex  capitalgain  capitalloss  hoursperweek  native  Salary  
0    1           24            0            39      37       0  
1    1            0            0            12      37       0  
2    1            0            0            39      37       0  
3    1            0            0            39      37       0  
4    0            0            0            39       4       0

In [14]:
test1 = test1.apply(LabelEncoder().fit_transform)
test1.head()

age  workclass  education  maritalstatus  occupation  relationship  race  \
0    8          2          1              4           6             3     2   
1   21          2         11              2           4             0     4   
2   11          1          7              2          10             0     4   
3   27          2         15              2           6             0     2   
4   17          2          0              4           7             1     4   

   sex  capitalgain  capitalloss  hoursperweek  native  Salary  
0    1            0            0            39      37       0  
1    1            0            0            49      37       0  
2    1            0            0            39      37       1  
3    1           87            0            39      37       1  
4    1            0            0            29      37       0

In [15]:
train1.shape

(30161, 13)

In [16]:
x_train = train1.iloc[:,:-1]
y_train = train1.iloc[:,-1]
print(x_train.head())
y_train.head()

   age  workclass  education  maritalstatus  occupation  relationship  race  \
0   22          5          9              4           0             1     4   
1   33          4          9              2           3             0     4   
2   21          2         11              0           5             1     4   
3   36          2          1              2           5             0     2   
4   11          2          9              2           9             5     2   

   sex  capitalgain  capitalloss  hoursperweek  native  
0    1           24            0            39      37  
1    1            0            0            12      37  
2    1            0            0            39      37  
3    1            0            0            39      37  
4    0            0            0            39       4  


0    0
1    0
2    0
3    0
4    0
Name: Salary, dtype: int32

In [17]:
x_test = test1.iloc[:,:-1]
y_test = test1.iloc[:,-1]
print(x_test.head())
y_test.head()

   age  workclass  education  maritalstatus  occupation  relationship  race  \
0    8          2          1              4           6             3     2   
1   21          2         11              2           4             0     4   
2   11          1          7              2          10             0     4   
3   27          2         15              2           6             0     2   
4   17          2          0              4           7             1     4   

   sex  capitalgain  capitalloss  hoursperweek  native  
0    1            0            0            39      37  
1    1            0            0            49      37  
2    1            0            0            39      37  
3    1           87            0            39      37  
4    1            0            0            29      37  


0    0
1    0
2    1
3    1
4    0
Name: Salary, dtype: int32

### Naive Bayes Model Building with Multinomial

In [21]:
model_mb = MultinomialNB()
model_mb.fit(x_train, y_train)

MultinomialNB()

In [22]:
preds_mb = model_mb.predict(x_test)

In [25]:
accuracy_mb = np.mean(preds_mb == y_test)

In [36]:
print('accuracy with Multinomial is ', accuracy_mb*100)

accuracy with Multinomial is  78.01460823373174


### Naive Bayes Model Building with Gaussian

In [28]:
model_gb = GaussianNB()
model_gb.fit(x_train, y_train)

GaussianNB()

In [29]:
preds_gb = model_gb.predict(x_test)

In [30]:
accuracy_gb = np.mean(preds_gb == y_test)

In [37]:
print('accuracy with Gaussian is ', accuracy_gb*100)

accuracy with Gaussian is  79.30942895086321


### Naive Bayes Model Building with Bernoulli

In [32]:
model_bb = BernoulliNB()
model_bb.fit(x_train, y_train)

BernoulliNB()

In [33]:
preds_bb = model_bb.predict(x_test)

In [34]:
accuracy_bb = np.mean(preds_bb == y_test)

In [38]:
print('accuracy with BernoulliNB is ', accuracy_bb*100)

accuracy with BernoulliNB is  72.84196547144755


First I have load both train and test datasets. Then i have check for null values and data types. In this both datasets there is no null values. In both train and test data education and educationno columns are same so ihave drop the educationno in both train and test. After that by using labelencoder method i have converted all categorical to numeric. Then i have divided train data to xtrain and ytrain. And test data to xtest and ytest. I have build naive bayes model with Multinomial method then fit xtrain ytrain on model. ihave done model prediction with xtest and i got model accuracy 78%. Then i have build model with Gaussian method i got model accuracy is 79% and with Bernoulli method i got model accuracy is 72%.